pkg/diagnostics/networkpod/pod.go

```diff
@@ -0,0 +1,206 @@
+		return false, errors.New("must have kube client")
+	} else if d.OSClient == nil {
+		return false, errors.New("must have openshift client")
+	}
+	return true, nil
+}
+
+// Check is part of the Diagnostic interface; it runs the actual diagnostic logic
+func (d CheckPodNetwork) Check() types.DiagnosticResult {
+	r := types.NewDiagnosticResult(CheckPodNetworkName)
+
+	pluginName, ok, err := util.GetOpenShiftNetworkPlugin(d.OSClient)
+	if err != nil {
+		r.Error("DPodNet1001", err, fmt.Sprintf("Checking network plugin failed. Error: %s", err))
+		return r
+	}
+	if !ok {
+		r.Warn("DPodNet1002", nil, fmt.Sprintf("Skipping pod connectivity test. Reason: Not using openshift network plugin."))
+		return r
+	}
+
+	localPods, nonlocalPods, err := util.GetLocalAndNonLocalPods(d.KubeClient)
+	if err != nil {
+		r.Error("DPodNet1003", err, fmt.Sprintf("Getting local and nonlocal pods failed. Error: %s", err))
+		return r
+	}
+
+	var vnidMap map[string]uint32
+	if sdnplugin.IsOpenShiftMultitenantNetworkPlugin(pluginName) {
+		netnsList, err := d.OSClient.NetNamespaces().List(kapi.ListOptions{})
+		if err != nil {
+			r.Error("DPodNet1004", err, fmt.Sprintf("Getting all network namespaces failed. Error: %s", err))
+			return r
+		}
+
+		vnidMap = map[string]uint32{}
+		for _, netns := range netnsList.Items {
+			vnidMap[netns.NetName] = netns.NetID
+		}
+	}
+
+	localGlobalPods, localNonGlobalPods := getGlobalAndNonGlobalPods(localPods, vnidMap)
+	nonlocalGlobalPods, nonlocalNonGlobalPods := getGlobalAndNonGlobalPods(nonlocalPods, vnidMap)
+
+	checkSameNodePodToPodConnection(localGlobalPods, localNonGlobalPods, vnidMap, r)
+	checkDifferentNodePodToPodConnection(localGlobalPods, localNonGlobalPods, nonlocalGlobalPods, nonlocalNonGlobalPods, vnidMap, r)
+	return r
+}
+
+func getGlobalAndNonGlobalPods(pods []kapi.Pod, vmap map[string]uint32) ([]kapi.Pod, []kapi.Pod) {
+	globalPods := []kapi.Pod{}
+	nonGlobalPods := []kapi.Pod{}
+
+	for _, pod := range pods {
+		if (vmap == nil) || (vmap != nil && vmap[pod.Namespace] == api.GlobalVNID) {
+			globalPods = append(globalPods, pod)
+		} else {
+			nonGlobalPods = append(nonGlobalPods, pod)
+		}
+	}
+	return globalPods, nonGlobalPods
+}
+func checkDifferentNodePodToPodConnection(localGlobalPods, localNonGlobalPods, nonlocalGlobalPods, nonlocalNonGlobalPods []kapi.Pod, vmap map[string]uint32, r types.DiagnosticResult) {
+	// Applicable to flat and multitenant networks
+	if len(localGlobalPods) > 0 && len(nonlocalGlobalPods) > 0 {
+		checkPodToPodConnection(localGlobalPods[0], nonlocalGlobalPods[0], vmap, r)
+	} else {
+		r.Warn("DPodNet1005", nil, fmt.Sprintf("Skipping pod connectivity test for global projects on different nodes. Reason: Couldn't find 2 global pods."))
+	}
+
+	// Applicable to multitenant network
+	isMultitenant := (vmap != nil)
+	if isMultitenant {
+		if len(localNonGlobalPods) > 0 && len(nonlocalNonGlobalPods) > 0 {
+			checkPodToPodConnection(localNonGlobalPods[0], nonlocalNonGlobalPods[0], vmap, r)
```

Comment:
The multitenant tests (both here and in checkSameNodePodToPodConnection()) probably need to make more effort to test something useful. It's just luck here whether we end up testing (a) two pods in the same namespace, (b) two pods in different isolated namespaces, (c) two pods in different-but-joined namespaces